File: internal/invite/invite.go
// Package invite encodes and parses Chimera share URLs so a client can be
// configured without typing seed/PSK by hand.
//
// Canonical form:
//
//	chimera://v1/<base64url({"v":1,"a":"host:port","s":"<64 hex>","p":"<64 hex>","g":0,"n":"optional"})>
//
// Also accepted: chimera://connect?addr=&seed=&psk=&generation=&name=,
// a pasted client.json object, or the same URL buried in surrounding text.
// The URL is the secret; callers must not log it.
package invite

import (
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
	"strings"
)

const schemePrefix = "chimera://v1/"

// Profile is one client endpoint. Hex fields are lowercase 64-char strings.
type Profile struct {
	Addr       string
	SeedHex    string
	PSKHex     string
	Generation uint64
	Name       string
}

type wire struct {
	V    int    `json:"v"`
	Addr string `json:"a"`
	Seed string `json:"s"`
	PSK  string `json:"p"`
	Gen  uint64 `json:"g"`
	Name string `json:"n,omitempty"`
}

// Format returns a chimera://v1/… URL. Name is optional.
func Format(p Profile) (string, error) {
	norm, err := normalize(p)
	if err != nil {
		return "", err
	}
	raw, err := json.Marshal(wire{
		V:    1,
		Addr: norm.Addr,
		Seed: norm.SeedHex,
		PSK:  norm.PSKHex,
		Gen:  norm.Generation,
		Name: norm.Name,
	})
	if err != nil {
		return "", err
	}
	return schemePrefix + base64.RawURLEncoding.EncodeToString(raw), nil
}

// Parse accepts an invite URL, query URL, or client.json text.
func Parse(text string) (Profile, error) {
	s := strings.TrimSpace(text)
	if s == "" {
		return Profile{}, fmt.Errorf("empty invite")
	}
	if extracted := extractURL(s); extracted != "" {
		s = extracted
	}
	switch {
	case strings.HasPrefix(s, "{"):
		return parseJSON(s)
	case strings.HasPrefix(s, "chimera://v1/"):
		return parseV1(strings.TrimPrefix(s, "chimera://v1/"))
	case strings.HasPrefix(s, "chimera://connect"):
		return parseConnect(s)
	default:
		return Profile{}, fmt.Errorf("not a chimera invite")
	}
}

func parseV1(payload string) (Profile, error) {
	payload = strings.TrimSpace(payload)
	payload = strings.TrimRight(payload, "/")
	raw, err := base64.RawURLEncoding.DecodeString(payload)
	if err != nil {
		// Some messengers re-pad; try with padding.
		raw, err = base64.URLEncoding.DecodeString(payload)
		if err != nil {
			return Profile{}, fmt.Errorf("invite payload: %w", err)
		}
	}
	var w wire
	if err := json.Unmarshal(raw, &w); err != nil {
		return Profile{}, fmt.Errorf("invite json: %w", err)
	}
	if w.V != 0 && w.V != 1 {
		return Profile{}, fmt.Errorf("unsupported invite version %d", w.V)
	}
	return normalize(Profile{
		Addr:       w.Addr,
		SeedHex:    w.Seed,
		PSKHex:     w.PSK,
		Generation: w.Gen,
		Name:       w.Name,
	})
}

func parseConnect(rawURL string) (Profile, error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return Profile{}, err
	}
	q := u.Query()
	gen := uint64(0)
	if g := first(q, "generation", "g"); g != "" {
		n, err := strconv.ParseUint(g, 10, 64)
		if err != nil {
			return Profile{}, fmt.Errorf("generation: %w", err)
		}
		gen = n
	}
	return normalize(Profile{
		Addr:       first(q, "addr", "server", "serverAddr"),
		SeedHex:    first(q, "seed", "seedHex"),
		PSKHex:     first(q, "psk", "pskHex"),
		Generation: gen,
		Name:       first(q, "name", "n"),
	})
}

func parseJSON(s string) (Profile, error) {
	var m map[string]any
	if err := json.Unmarshal([]byte(s), &m); err != nil {
		return Profile{}, fmt.Errorf("json: %w", err)
	}
	gen := uint64(0)
	switch v := m["generation"].(type) {
	case float64:
		if v < 0 {
			return Profile{}, fmt.Errorf("generation must be >= 0")
		}
		gen = uint64(v)
	case string:
		if v != "" {
			n, err := strconv.ParseUint(v, 10, 64)
			if err != nil {
				return Profile{}, fmt.Errorf("generation: %w", err)
			}
			gen = n
		}
	}
	return normalize(Profile{
		Addr:       jsonString(m, "serverAddr", "a", "addr"),
		SeedHex:    jsonString(m, "seedHex", "s", "seed_hex"),
		PSKHex:     jsonString(m, "pskHex", "p", "psk_hex"),
		Generation: gen,
		Name:       jsonString(m, "name", "n"),
	})
}

func normalize(p Profile) (Profile, error) {
	p.Addr = strings.TrimSpace(p.Addr)
	p.Name = strings.TrimSpace(p.Name)
	seed, err := normHex("seed", p.SeedHex)
	if err != nil {
		return Profile{}, err
	}
	psk, err := normHex("psk", p.PSKHex)
	if err != nil {
		return Profile{}, err
	}
	if p.Addr == "" {
		return Profile{}, fmt.Errorf("server address is empty")
	}
	p.SeedHex = seed
	p.PSKHex = psk
	return p, nil
}

func normHex(label, s string) (string, error) {
	s = strings.TrimSpace(strings.ToLower(s))
	s = strings.TrimPrefix(s, "0x")
	if len(s) != 64 {
		return "", fmt.Errorf("%s must be 64 hex characters", label)
	}
	if _, err := hex.DecodeString(s); err != nil {
		return "", fmt.Errorf("%s: %w", label, err)
	}
	return s, nil
}

func extractURL(s string) string {
	for _, prefix := range []string{"chimera://v1/", "chimera://connect"} {
		i := strings.Index(s, prefix)
		if i < 0 {
			continue
		}
		rest := s[i:]
		if cut := strings.IndexAny(rest, " \t\r\n<>\"'"); cut >= 0 {
			rest = rest[:cut]
		}
		return rest
	}
	return ""
}

func first(q url.Values, keys ...string) string {
	for _, k := range keys {
		if v := strings.TrimSpace(q.Get(k)); v != "" {
			return v
		}
	}
	return ""
}

func jsonString(m map[string]any, keys ...string) string {
	for _, k := range keys {
		switch v := m[k].(type) {
		case string:
			if strings.TrimSpace(v) != "" {
				return strings.TrimSpace(v)
			}
		}
	}
	return ""
}

File: apps/android/app/src/main/java/com/chimera/vpn/Invite.kt
package com.chimera.vpn

// Pure-JVM by design: no android.* imports, so Invites is unit-testable
// with plain JUnit (see app/src/test). Base64 uses java.util.Base64
// (available since API 26) and connect-URL queries are parsed manually.

import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.util.Base64
import java.util.Locale

/**
 * Chimera share URLs. Keep in lockstep with Go `internal/invite`.
 *
 * Canonical: chimera://v1/<base64url({"v":1,"a":"host:port","s":"<64 hex>","p":"<64 hex>","g":0})>
 */
data class Invite(
    val addr: String,
    val seedHex: String,
    val pskHex: String,
    val generation: Long,
    val name: String = ""
)

object Invites {
    private const val V1 = "chimera://v1/"
    private const val CONNECT = "chimera://connect"
    private val B64ENC: Base64.Encoder = Base64.getUrlEncoder().withoutPadding()
    private val B64DEC: Base64.Decoder = Base64.getUrlDecoder()

    fun format(p: Invite): String {
        val n = normalize(p)
        val obj = JSONObject()
            .put("v", 1)
            .put("a", n.addr)
            .put("s", n.seedHex)
            .put("p", n.pskHex)
            .put("g", n.generation)
        if (n.name.isNotEmpty()) obj.put("n", n.name)
        val raw = obj.toString().toByteArray(StandardCharsets.UTF_8)
        return V1 + B64ENC.encodeToString(raw)
    }

    fun parse(text: String): Invite {
        var s = text.trim()
        if (s.isEmpty()) throw IllegalArgumentException("empty invite")
        extractUrl(s)?.let { s = it }
        return when {
            s.startsWith("{") -> parseJson(s)
            s.startsWith(V1) -> parseV1(s.removePrefix(V1).trim().trimEnd('/'))
            s.startsWith(CONNECT) -> parseConnect(s)
            else -> throw IllegalArgumentException("not a chimera invite")
        }
    }

    private fun parseV1(payload: String): Invite {
        // java.util.Base64 tolerates missing padding, which covers both the
        // NO_PADDING canonical form and padded variants.
        val decoded = B64DEC.decode(payload)
        val obj = JSONObject(String(decoded, StandardCharsets.UTF_8))
        val v = obj.optInt("v", 1)
        if (v != 0 && v != 1) throw IllegalArgumentException("unsupported invite version $v")
        return normalize(
            Invite(
                addr = obj.optString("a"),
                seedHex = obj.optString("s"),
                pskHex = obj.optString("p"),
                generation = obj.optLong("g", 0L),
                name = obj.optString("n")
            )
        )
    }

    private fun parseConnect(raw: String): Invite {
        val params = queryParams(raw)
        val gen = firstQuery(params, "generation", "g")
        return normalize(
            Invite(
                addr = firstQuery(params, "addr", "server", "serverAddr"),
                seedHex = firstQuery(params, "seed", "seedHex"),
                pskHex = firstQuery(params, "psk", "pskHex"),
                generation = gen.toLongOrNull() ?: 0L,
                name = firstQuery(params, "name", "n")
            )
        )
    }

    private fun queryParams(raw: String): Map<String, String> {
        val query = raw.substringAfter('?', "")
        val out = LinkedHashMap<String, String>()
        for (pair in query.split('&')) {
            if (pair.isEmpty()) continue
            val i = pair.indexOf('=')
            val k = if (i >= 0) pair.substring(0, i) else pair
            val v = if (i >= 0) pair.substring(i + 1) else ""
            if (k.isEmpty() || out.containsKey(k)) continue
            out[k] = java.net.URLDecoder.decode(v, StandardCharsets.UTF_8)
        }
        return out
    }

    private fun parseJson(s: String): Invite {
        val obj = JSONObject(s)
        return normalize(
            Invite(
                addr = firstJson(obj, "serverAddr", "a", "addr"),
                seedHex = firstJson(obj, "seedHex", "s", "seed_hex"),
                pskHex = firstJson(obj, "pskHex", "p", "psk_hex"),
                generation = when {
                    obj.has("generation") && !obj.isNull("generation") ->
                        obj.optLong("generation", 0L)
                    else -> 0L
                },
                name = firstJson(obj, "name", "n")
            )
        )
    }

    private fun normalize(p: Invite): Invite {
        val seed = normHex("seed", p.seedHex)
        val psk = normHex("psk", p.pskHex)
        val addr = p.addr.trim()
        if (addr.isEmpty()) throw IllegalArgumentException("server address is empty")
        return Invite(addr, seed, psk, p.generation, p.name.trim())
    }

    private fun normHex(label: String, raw: String): String {
        var s = raw.trim().lowercase(Locale.US)
        if (s.startsWith("0x")) s = s.substring(2)
        if (s.length != 64 || !s.all { it in '0'..'9' || it in 'a'..'f' }) {
            throw IllegalArgumentException("$label must be 64 hex characters")
        }
        return s
    }

    private fun extractUrl(s: String): String? {
        for (prefix in listOf(V1, CONNECT)) {
            val i = s.indexOf(prefix)
            if (i < 0) continue
            var rest = s.substring(i)
            val cut = rest.indexOfFirst { it in " \t\r\n<>\"'" }
            if (cut >= 0) rest = rest.substring(0, cut)
            return rest
        }
        return null
    }

    private fun firstQuery(params: Map<String, String>, vararg keys: String): String {
        for (k in keys) {
            val v = params[k]?.trim().orEmpty()
            if (v.isNotEmpty()) return v
        }
        return ""
    }

    private fun firstJson(obj: JSONObject, vararg keys: String): String {
        for (k in keys) {
            val v = obj.optString(k).trim()
            if (v.isNotEmpty()) return v
        }
        return ""
    }
}
